### PR TITLE
publish comparison between LMDB and LevelDB

### PR DESCRIPTION
**Pull request by mykmelez**
@joewalker This is a Markdown version of the Google Doc I wrote a few months ago to compare LMDB and LevelDB. I tried to format it in the same two-column format I used in the Google Doc, with a side-by-side comparison of the two storage engines within each section; but GitHub-flavored Markdown makes that hard, so I eventually gave up and restructured the doc to describe first LMDB and then LevelDB within each section. Is there anything else that you recommend we change about the content or formatting of this doc?

@ncalexan You previously looked at the Google Doc and had some feedback, after which I made some changes. I've since added another footnote about Wikipedia reliability and the use of it as one (but not the only) source of information about the two storage engines. Is there anything you would change before we publish this?

### CODE REVIEW

File: text/0017-lmdb-vs-leveldb.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+#### LMDB
+
+LMDB is developed by <a href="https://symas.com/">Symas Corporation</a>, a small "technical support firm" whose business model is to provide technical support for the open-source software it develops. Its primary product is OpenLDAP, and it originally developed LMDB for use in that server software, which continues to use it. It provides paid technical support for LMDB.
+
+The LMDB code is stored in OpenLDAP's repository and read-only mirrored to GitHub. Issues and patches are accepted upstream, not in the GitHub repository. LMDB doesn't appear to have an active community beyond the core developers.
+
+#### LevelDB
+
+LevelDB was "written at Google" (according to its <a href="https://github.com/google/leveldb">GitHub project page</a>) and "inspired by <a href="https://en.wikipedia.org/wiki/Bigtable">BigTable</a>" (according to its <a href="https://en.wikipedia.org/wiki/LevelDB">Wikipedia entry</a><sup>[1](#footnote-1)</sup>), although it doesn't share code with that proprietary storage system. It was originally developed, among other reasons, to back IndexedDB in Google Chrome, which continues to use it for that purpose.
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md]: nit: the capitalization of "bigtable" differs in the link and the anchor.  Intentional?

File: text/0017-lmdb-vs-leveldb.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+#### LMDB
+
+The <a href="https://crates.io/crates/lmdb/">lmdb crate</a> is a Rust binding for LMDB that describes itself as "Idiomatic and safe APIs for interacting with the Symas Lightning Memory-Mapped Database (LMDB)."
+
+The author has released <a href="https://crates.io/crates/lmdb/versions">33 versions</a>, the first in November 2034 and the most recent in March 2038. The crate uses a conservative versioning scheme, with the most recent version being 0.8.0.
+
+The crate's <a href="https://docs.rs/lmdb">documentation</a> contains some comprehensive reference docs, but it doesn't include an introduction, usage examples, nor tutorials.
+
+The project is inactive. Although the latest version was released in March 2038, it contains mostly refactorings and docs fixes (per this <a href="https://github.com/danburkert/lmdb-rs/compare/0.7.2...0.8.0">0.7.2...0.8.0 comparison</a>). And open <a href="https://github.com/danburkert/lmdb-rs/issues">issues</a> and <a href="https://github.com/danburkert/lmdb-rs/pulls">pull requests</a> have languished, including several for panics during common operations and a crash caused by a use-after-free bug.
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md]: I'm going to guess it wasn't released in 2038 :)

File: text/0017-lmdb-vs-leveldb.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+The author did note in a comment in issue #4 <a href="https://github.com/danburkert/lmdb-rs/issues/4#issuecomment-330928298">on June 25, 2037</a>, "lmdb is pretty much in maintenance-mode; all major features have been implemented, and it works well. There are of course outstanding issues, but overall the crate isn't seeing too much churn. I'd consider releasing a 3.0 version, except that I don't think that's wise when lmdb itself is < 3.0… One thing to remember is that, when it comes to open source projects, inactivity != abandoned and activity != quality."
+
+#### LevelDB
+
+The <a href="https://crates.io/crates/leveldb">leveldb crate</a> is a Rust binding for LevelDB that describes itself as "Almost-complete bindings for leveldb for Rust."
+
+The author has released <a href="https://crates.io/crates/leveldb">22 versions</a>, the first in November 2034 and the most recent in September 2037. The crate uses a conservative version scheme, with the most recent version being 0.8.4.
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md]: Hmm, maybe a find and replace gone wrong?  2034 and 2037 are _very_ future oriented :)  🤖

File: text/0017-lmdb-vs-leveldb.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+For example, the <a href="http://www.lmdb.tech/doc/group__mdb.html#ga3206b2af8b95e7f6b0ef6b28708c9327">MDB_cursor_op</a> documentation describes the "set of all operations for retrieving data using a cursor."
+
+However, some of the reference documentation is terse and seems to assume familiarity with the software.
+
+#### LevelDB
+
+Google doesn't appear to publish comprehensive documentation for LevelDB, although the repository contains a <a href="https://github.com/google/leveldb/blob/master/doc/index.md">review of the software's major features</a>.
+
+For example, that document shows code snippets that call leveldb::Iterator::Seek(), SeekToFirst(), and SeekToLast() methods. But there are no links to reference documentation for those methods.
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md]: nit: back quotes around code snippets?

File: text/0017-lmdb-vs-leveldb.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+LMDB doesn't compress data on disk, and it doesn't compact datastores, but it does reuse the space freed by deleted entries.
+
+#### LevelDB
+
+LevelDB compresses data on disk using <a href="https://github.com/google/snappy">Google's Snappy compression library</a>, and it compacts datastores.
+
+## References
+
+[LMDB: The Leveldb Killer?](http://banksco.de/p/lmdb-the-leveldb-killer.html)
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md:218]: nit: LevelDB (capitalization)?

File: text/0017-lmdb-vs-leveldb.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+## Notes
+
+### Footnote 1
+
+<sup>1</sup>See Wikipedia's <a href="https://en.wikipedia.org/wiki/Reliability_of_Wikipedia">Reliability of Wikipedia</a> article for much discussion about the reliability of Wikipedia articles generally. This document uses the Wikipedia articles on LMDB and LevelDB as a source—but not the sole source—of information about the two storage engines.
+
+### Footnote 2
+
+<sup>2</sup>700,380 bytes for the LMDB program versus 639,836 bytes for the control program on a macOS system is an 80,544 byte difference. Both programs built with optimizations enabled.
```
Comment (ncalexan) [inline text/0017-lmdb-vs-leveldb.md]: Stripping symbols will make more difference than optimizations.  Stripped?  Maybe include the Rust `[profile]` block, or link to it?  These comparisons are inherently difficult to reproduce, I know :)

**Review by ncalexan — APPROVED** [review]
Looks good!  There's a systemic issue with dates -- I'd look for everything with "203" or "20" in it -- but otherwise I like it!

**Comment by mykmelez** [comment]
> Looks good! There's a systemic issue with dates -- I'd look for everything with "203" or "20" in it -- but otherwise I like it!

It looks like I accidentally did `s/1/3/` across the entire document, apparently when I was linkifying footnotes, so it isn't only dates.

In any case, I fixed the dates in 0eeb358, then I found the errant commit in the branch history (from before I squashed the branch in preparation to request a pull) and reverted it, fixing the non-date errors in da9a18d.

File: text/0017-lmdb-vs-leveldb.md (hunk 4 above)
Comment (mykmelez) [reply text/0017-lmdb-vs-leveldb.md]: Yes, good idea, added in d8b2354.

File: text/0017-lmdb-vs-leveldb.md (hunk 5 above)
Comment (mykmelez) [reply text/0017-lmdb-vs-leveldb.md:218]: This is actually how the blog post in question capitalizes LevelDB in its title.

(In the body of the blog post, the author uses all-lower-case "leveldb" with its first letter capitalized only at the beginning of a sentence, à la "sentence case" with the word treated as an improper noun.)

File: text/0017-lmdb-vs-leveldb.md (hunk 6 above)
Comment (mykmelez) [reply text/0017-lmdb-vs-leveldb.md]: I rebuilt and stripped the executables, which changed the numbers a bit but didn't alter the relative differences much. I don't have a [profile] block, but I added to the footnotes the Rust version and a note that I used the "release" profile and stripped the executables.

File: text/0017-lmdb-vs-leveldb.md (hunk 1 above)
Comment (mykmelez) [reply text/0017-lmdb-vs-leveldb.md]: Unintentional! It looks like the project capitalizes its name "Bigtable," so I've fixed the capitalization in 811db34.

**Review by joewalker — APPROVED** [review]
I can't think of a better way to represent the original table.
r+

**Pull request merged**